commit f19cdf7aea6616988e3723f5e185d3322b58fd1f (refs/changes/51/5051/6)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-11-30T09:53:17-08:00 (10 months ago)
    
    OS-7361 SmartOS initial set up password prompt mangles some patterns
    Reviewed by: Todd Whiteman <todd.whiteman@joyent.com>
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/scripts/prompt-config.sh b/scripts/prompt-config.sh
index fa1258f2..c34db39f 100755
--- a/scripts/prompt-config.sh
+++ b/scripts/prompt-config.sh
@@ -640,13 +640,19 @@ promptnic()
     val=$mac_addr
 }
 
+readpw()
+{
+	IFS='' read -r -s pw
+	printf '%s' "$pw"
+}
+
 promptpw()
 {
 	def="$3"
     key="$4"
 
 	if [[ -n ${key} ]]; then
-		preset_val=$(getanswer "${key}")
+		preset_val="$(getanswer "${key}")"
 	fi
 
 	trap "" SIGINT
@@ -654,7 +660,7 @@ promptpw()
 		val=""
 		while [ -z "$val" ]; do
 			if [[ -n ${preset_val} ]]; then
-				val=${preset_val}
+				val="${preset_val}"
 			else
 				if [ -z "$def" ]; then
 					printf "%s: " "$1"
@@ -662,9 +668,7 @@ promptpw()
 					printf "%s [enter to keep existing]: " \
 					    "$1"
 				fi
-				stty -echo
-				read val
-				stty echo
+				val="$(readpw)"
 				echo
 			fi
 			if [ -n "$val" ]; then
@@ -686,7 +690,7 @@ promptpw()
 				fi
 			else
 				if [ -n "$def" ]; then
-					val=$def
+					val="$def"
 					return
 				else
 					echo "A value must be provided."
@@ -697,12 +701,10 @@ promptpw()
 		cval=""
 		while [ -z "$cval" ]; do
 			if [[ -n ${preset_val} ]]; then
-				cval=${preset_val}
+				cval="${preset_val}"
 			else
 				printf "%s: " "Confirm password"
-				stty -echo
-				read cval
-				stty echo
+				cval="$(readpw)"
 				echo
 			fi
 			[ -n "$cval" ] && break
